main/src/com/google/refine/expr/functions/arrays/Zip.java

```diff
@@ -0,0 +1,115 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+package com.google.refine.expr.functions.arrays;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import java.util.Spliterator;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+import com.google.refine.expr.EvalError;
+import com.google.refine.grel.ControlFunctionRegistry;
+import com.google.refine.grel.EvalErrorMessage;
+import com.google.refine.grel.Function;
+import com.google.refine.grel.FunctionDescription;
+
+public class Zip implements Function {
+
+    @Override
+    public Object call(Properties bindings, Object[] args) {
+        if (args.length >= 2) {
+            List<Spliterator> iterators = new ArrayList<>();
+            for (Object v : args) {
+                if (v == null || !(v.getClass().isArray() || v instanceof List<?> || v instanceof ArrayNode)) {
+                    return new EvalError(
+                            EvalErrorMessage.expects_at_least_two_or_more_array_args(ControlFunctionRegistry.getFunctionName(this)));
+                } else if (v.getClass().isArray()) {
+                    iterators.add(Arrays.stream((Object[]) v).spliterator());
+                } else if (v instanceof ArrayNode) {
+                    ArrayNode a = (ArrayNode) v;
+                    List<Object> ziparg = new ArrayList<>();
+                    if (a.isArray()) {
+                        for (JsonNode objNode : a) {
+                            ziparg.add(objNode);
+                        }
+                    }
+                    iterators.add(ziparg.stream().spliterator());
+                } else {
+                    iterators.add(((List<Object>) v).spliterator());
+                }
+            }
+
+            int minSize = Integer.MAX_VALUE;
+            for (Spliterator<?> iterator : iterators) {
+                minSize = Math.min(minSize, (int) iterator.estimateSize());
+            }
+
+            List<List> output = new ArrayList<>();
+            List<Object> currentElements = new ArrayList<>(minSize);
+
+            for (int i = 0; i < minSize; i++) {
+                currentElements.clear();
+                for (Spliterator<?> iterator : iterators) {
+                    Object[] elementHolder = new Object[1];
+                    if (iterator.tryAdvance(e -> elementHolder[0] = e)) {
+                        currentElements.add(elementHolder[0]);
+                    }
+                }
+                output.add(new ArrayList<>(currentElements));
```

Comment:
```suggestion
                    if (!iterator.tryAdvance(e -> currentElements.add(e)) {
                        done = true;
                    }
                }
                if (!done) {
                    output.add(currentElements);
                }
```

You can add the element to your output list directly inside the tryAdvance method and by inverting the if test, you can set a `done` flag when one of your iterators is empty and it's time to leave.